File: matlab/derand1bin.m
function [xmin, fmin] = derand1bin(fitfun, lb, ub, maxfunevals)
% Algorithmic parameters
D	= numel(lb);
NP	= 5 * D;
F	= 0.7;
CR	= 0.5;

% Initialize population (4)
X = zeros(D, NP);
for i = 1 : NP
	X(:, i) = lb + (ub - lb) .* rand(D, 1);
end

% Initialize variables
V = X;
U = X;

% Evaluation
counteval = 0;
f = zeros(1, NP);
for i = 1 : NP
	f(i) = feval(fitfun, X(:, i));
end
counteval = counteval + NP;

while counteval <= maxfunevals - NP	
	% Mutation (5)
	for i = 1 : NP
		r1 = floor(1 + NP * rand);
		r2 = floor(1 + NP * rand);
		r3 = r2;		
		while r2 == r3
			r3 = floor(1 + NP * rand);
		end
		V(:, i) = X(:, r1) + F * (X(:, r2) - X(:, r3));
	end
	
	% Crossover (10)
	for i = 1 : NP
		U(:, i) = bincrossover(X(:, i), V(:, i), U(:, i), CR, D);
	end
	
	% Selection	(11)
	for i = 1 : NP
		fui = feval(fitfun, U(:, i));
		counteval = counteval + 1;
		
		if fui < f(i)
			f(i) = fui;
			X(:, i) = U(:, i);
		end
	end
end

[fmin, fminidx] = min(f);
xmin = X(:, fminidx);
end

File: cpp/DE-EIG/DE-EIG/derand1bin.cpp
#include "stdafx.h"
using std::vector;

void derand1bin(vector<double> &xmin, double &fmin, double lb[], double ub[], int maxfunevals, int D)
{
	// Algorithmic parameters
	int				NP	= 5 * D;
	double			F	= 0.7;
	vector<double>	CR;

	for (int i = 0; i < NP; ++i)
	{
		CR.push_back(0.5);
	}

	// Initialize population (4)
	vector<vector<double>> X;
	for (int i = 0; i < NP; ++i)
	{
		vector<double> xi;
		for (int j = 0; j < D; ++j)
		{
			double xij = lb[j] + (ub[j] - lb[j]) * ((double) rand() / RAND_MAX);
			xi.push_back(xij);
		}
		X.push_back(xi);
	}

	vector<vector<double>> V = X;
	vector<vector<double>> U = X;
	
	// Evaluation
	int counteval = 0;
	vector<double> f;
	for (int i = 0; i < NP; ++i)
	{
		f.push_back(ellipsoidalrot(X[i].data()));
	}
	counteval = counteval + NP;

	while (counteval <= maxfunevals - NP)
	{
		// Mutation (5)
		for (int i = 0; i < NP; ++i)
		{
			int r1 = rand() % NP;
			int r2 = rand() % NP;
			int r3 = rand() % NP;
			while (r2 == r3)
			{
				r3 = rand() % NP;
			}
			for (int j = 0; j < D; ++j)
			{
				V[i][j] = X[r1][j] + F * (X[r2][j] - X[r3][j]);
			}
		}
		
		// Crossover (10)
		for (int i = 0; i < NP; ++i)
		{
			bincrossover(U[i], X[i], V[i], CR[i], D);
		}

		// Selection (11)
		for (int i = 0; i < NP; ++i)
		{
			double fui = ellipsoidalrot(U[i].data());
			if (fui < f[i])
			{
				f[i] = fui;
				for (int j = 0; j < D; ++j)
				{
					X[i][j] = U[i][j];
				}
			}
		}
		counteval += NP;
	}

	int fminidx = 0;
	fmin = f[0];
	for (int i = 1; i < NP; ++i) 
	{
		if (f[i] < fmin)
		{
			fmin = f[i];
			fminidx = i;
		}
	}
	xmin = X[fminidx];
}